README
= bcrypt-ruby

An easy way to keep your users' passwords secure.

== Why you should use bcrypt

If you store user passwords in the clear, then an attacker who steals a copy of your database has a giant list of emails
and passwords. Some of your users will only have one password -- for their email account, for their banking account, for
your application. A simple hack could escalate into massive identity theft.

It's your responsibility as a web developer to make your web application secure -- blaming your users for not being
security experts is not a professional response to risk.

bcrypt allows you to easily harden your application against these kinds of attacks.

== How to install bcrypt

  sudo gem install bcrypt-ruby
  
You'll need a working compiler. (Win32 folks should use Cygwin or um, something else.)

== How to use bcrypt in your Rails application

=== The +User+ model

  require 'bcrypt'
  
  class User < ActiveRecord::Base
    # users.password_hash in the database is a :string
    include BCrypt
    
    def password
      @password ||= Password.new(password_hash)
    end
    
    def password=(new_password)
      @password = Password.create(new_password)
      self.password_hash = @password
    end
    
  end

=== Creating an account

  def create
    @user = User.new(params[:user])
    @user.password = params[:password]
    @user.save!
  end

=== Authenticating a user

  def login
    @user = User.find_by_email(params[:email])
    if @user.password == params[:password]
      give_token
    else
      redirect_to home_url
    end
  end

=== If a user forgets their password?

  # assign them a random one and mail it to them, asking them to change it
  def forgot_password
    @user = User.find_by_email(params[:email])
    random_password = Array.new(10).map { (65 + rand(58)).chr }.join
    @user.password = random_password
    @user.save!
    Mailer.create_and_deliver_password_change(@user, random_password)
  end

== How to use bcrypt-ruby in general

  require 'bcrypt'
  
  my_password = BCrypt::Password.create("my password") #=> "$2a$10$vI8aWBnW3fID.ZQ4/zo1G.q1lRps.9cGLcZEiGDMVr5yUP1KUOYTa"
  
  my_password.version              #=> "2a"
  my_password.cost                 #=> 10
  my_password == "my password"     #=> true
  my_password == "not my password" #=> false
  
  my_password = BCrypt::Password.new("$2a$10$vI8aWBnW3fID.ZQ4/zo1G.q1lRps.9cGLcZEiGDMVr5yUP1KUOYTa")
  my_password == "my password"     #=> true
  my_password == "not my password" #=> false

Check the rdocs for more details -- BCrypt, BCrypt::Password.

== How bcrypt() works

bcrypt() is a hashing algorithm designed by Niels Provos and David Mazières of the OpenBSD Project.

=== Background

Hash algorithms take a chunk of data (e.g., your user's password) and create a "digital fingerprint," or hash, of it.
Because this process is not reversible, there's no way to go from the hash back to the password.

In other words:

  hash(p) #=> <unique gibberish>

You can store the hash and check it against a hash made of a potentially valid password:

  <unique gibberish> =? hash(just_entered_password)

=== Rainbow Tables

But even this has weaknesses -- attackers can just run lists of possible passwords through the same algorithm, store the
results in a big database, and then look up the passwords by their hash:

  PrecomputedPassword.find_by_hash(<unique gibberish>).password #=> "secret1"

=== Salts

The solution to this is to add a small chunk of random data -- called a salt -- to the password before it's hashed:

  hash(salt + p) #=> <really unique gibberish>

The salt is then stored along with the hash in the database, and used to check potentially valid passwords:

  <really unique gibberish> =? hash(salt + just_entered_password)
  
bcrypt-ruby automatically handles the storage and generation of these salts for you.

Adding a salt means that an attacker has to have a gigantic database for each unique salt -- for a salt made of 4
letters, that's 456,976 different databases. Pretty much no one has that much storage space, so attackers try a
different, slower method -- throw a list of potential passwords at each individual password:

  hash(salt + "aadvark") =? <really unique gibberish>
  hash(salt + "abacus")  =? <really unique gibberish>
  etc.
  
This is much slower than the big database approach, but most hash algorithms are pretty quick -- and therein lies the
problem. Hash algorithms aren't usually designed to be slow, they're designed to turn gigabytes of data into secure
fingerprints as quickly as possible. bcrypt(), though, is designed to be computationally expensive:

  Ten thousand iterations:
               user     system      total        real
  md5      0.070000   0.000000   0.070000 (  0.070415)
  bcrypt  22.230000   0.080000  22.310000 ( 22.493822)

If an attacker was using Ruby to check each password, they could check ~140,000 passwords a second with MD5 but only
~450 passwords a second with bcrypt().

=== Cost Factors

In addition, bcrypt() allows you to increase the amount of work required to hash a password as computers get faster. Old
passwords will still work fine, but new passwords can keep up with the times.

The default cost factor used by bcrypt-ruby is 10, which is fine for session-based authentication. If you are using a
stateless authentication architecture (e.g., HTTP Basic Auth), you will want to lower the cost factor to reduce your
server load and keep your request times down. This will lower the security provided you, but there are few alternatives.

== More Information

bcrypt() is currently used as the default password storage hash in OpenBSD, widely regarded as the most secure operating
system available. 


For a more technical explanation of the algorithm and its design criteria, please read Niels Provos and David Mazières'
Usenix99 paper:
http://www.usenix.org/events/usenix99/provos.html

If you'd like more down-to-earth advice regarding cryptography, I suggest reading <i>Practical Cryptography</i> by Niels
Ferguson and Bruce Schneier:
http://www.schneier.com/book-practical.html

= Etc

Author  :: Coda Hale <coda.hale@gmail.com>
Website :: http://blog.codahale.com


CHANGELOG
3.1.19 June 22 2023
  - Deprecate passing the third argument to `BCrypt::Engine.hash_secret` [GH #207 by @sergey-alekseev]
  - Add GC guards so the C compiler won't optimize out references [GH #270]

3.1.18 May 16 2022
  - Unlock GVL when calculating hashes and salts [GH #260]
  - Fix compilation warnings in `ext/mri/bcrypt_ext.c` [GH #261]

3.1.17 Mar 14 2022
- Fix regex in validators to use \A and \z instead of ^ and $ [GH #121]
- Truncate secrets greater than 72 bytes in hash_secret [GH #255]
- Assorted test and doc improvements

3.1.16 Sep 3 2020
  - Fix compilation on FreeBSD. [GH #234]

3.1.15 July 21 2020
  - Remove GVL optimization.  Apparently it breaks things [GH #230]

3.1.14 July 21 2020
  - Start calibration from the minimum cost supported by the algorithm [GH #206 by @sergey-alekseev]

3.1.13 May 31 2019
  - No longer include compiled binaries for Windows. See GH #173.
  - Update C and Java implementations to latest versions [GH #182 by @fonica]
  - Bump default cost to 12 [GH #181 by @bdewater]
  - Remove explicit support for Rubies 1.8 and 1.9
  - Define SKIP_GNU token when building extension (Fixes FreeBSD >= 12) [GH #189 by @adam12]

3.1.12 May 16 2018
  - Add support for Ruby 2.3, 2.4, and 2.5 in compiled Windows binaries
  - Fix compatibility with libxcrypt - Fixes hash errors in Fedora 28 and Ubuntu 20 [GH #164 by @besser82]

3.1.11 Mar 06 2016
  - Add support for Ruby 2.2 in compiled Windows binaries

3.1.10 Jan 28 2015
  - Fix issue with dumping a BCrypt::Password instance to YAML in Ruby 2.2 [GH #107 by @mattwildig]

3.1.9  Oct 23 2014
  - Rebuild corrupt binaries

3.1.8  Oct 23 2014
  - Add support for Ruby 2.1 in compiled Windows binaries [GH #102]

3.1.7  Feb 24 2014
  - Rebuild corrupt Java binary version of gem [GH #90]
  - The 2.1 support for Windows binaries alleged in 3.1.3 was a lie -- documentation removed

3.1.6  Feb 21 2014
  - Dummy version of "bcrypt-ruby" needed a couple version bumps to fix some
    bugs. It felt wrong to have that at a higher version than the real gem, so
    the real gem is getting bumped to 3.1.6.

3.1.3  Feb 21 2014
  - Add support for Ruby 2.1 in compiled Windows binaries
  - Rename gem from "bcrypt-ruby" to just "bcrypt". [GH #86 by @sferik]

3.1.2  Aug 26 2013
  - Add support for Ruby 1.8 and 2.0 (in addition to 1.9) in compiled Windows binaries
  - Add support for 64-bit Windows

3.1.1  Jul 10 2013
  - Remove support for Ruby 1.8 in compiled win32 binaries

3.1.0  May 07 2013
  - Add BCrypt::Password.valid_hash?(str) to check if a string is a valid bcrypt password hash
  - BCrypt::Password cost should be set to DEFAULT_COST if nil
  - Add BCrypt::Engine.cost attribute for getting/setting a default cost externally

3.0.1  Sep 12 2011
  - create raises an exception if the cost is higher than 31. GH #27

3.0.0  Aug 24 2011
  - Bcrypt C implementation replaced with a public domain implementation.
  - License changed to MIT

2.1.2  Sep 16 2009
  - Fixed support for Solaris, OpenSolaris.

2.1.1  Aug 14 2009
  - JVM 1.4/1.5 compatibility [Hongli Lai]

2.1.0  Aug 12 2009
  - Improved code coverage, unit tests, and build chain. [Hongli Lai]
  - Ruby 1.9 compatibility fixes. [Hongli Lai]
  - JRuby support, using Damien Miller's jBCrypt. [Hongli Lai]
  - Ruby 1.9 GIL releasing for high-cost hashes. [Hongli Lai]

2.0.5  Mar 11 2009
  - Fixed Ruby 1.8.5 compatibility. [Mike Pomraning]

2.0.4  Mar 09 2009
  - Added Ruby 1.9 compatibility. [Genki Takiuchi]
  - Fixed segfaults on some different types of empty strings. [Mike Pomraning]

2.0.3  May 07 2008
 - Made exception classes descend from StandardError, not Exception [Dan42]
 - Changed BCrypt::Engine.hash to BCrypt::Engine.hash_secret to avoid Merb
   sorting issues. [Lee Pope]

2.0.2  Jun 06 2007
 - Fixed example code in the README [Winson]
 - Fixed Solaris compatibility [Jeremy LaTrasse, Twitter crew]

2.0.1  Mar 09 2007
 - Fixed load path issues
 - Fixed crashes when hashing weird values (e.g., false, etc.)

2.0.0  Mar 07 2007
 - Removed BCrypt::Password#exactly_equals -- use BCrypt::Password#eql? instead.
 - Added BCrypt::Password#is_password?.
 - Refactored out BCrypt::Internals into more useful BCrypt::Engine.
 - Added validation of secrets -- nil is not healthy.

1.0.0  Feb 27 2007
 - Initial release.


ext/ffi/bcrypt.c
/*	$OpenBSD: bcrypt.c,v 1.22 2007/02/20 01:44:16 ray Exp $	*/

/* 
 * Modified by <coda.hale@gmail.com> on 2009-09-16:
 *
 *   - Standardized on stdint.h's numerical types and removed some debug cruft.
 *
 * Modified by <hongli@phusion.nl> on 2009-08-05:
 *
 *   - Got rid of the global variables; they're not thread-safe.
 *     Modified the functions to accept local buffers instead.
 *
 * Modified by <coda.hale@gmail.com> on 2007-02-27:
 *
 *   - Changed bcrypt_gensalt to accept a random seed as a parameter,
 *     to remove the code's dependency on arc4random(), which isn't
 *     available on Linux. 
 */

/*
 * Copyright 1997 Niels Provos <provos@physnet.uni-hamburg.de>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed by Niels Provos.
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* This password hashing algorithm was designed by David Mazieres
 * <dm@lcs.mit.edu> and works as follows:
 *
 * 1. state := InitState ()
 * 2. state := ExpandKey (state, salt, password) 3.
 * REPEAT rounds:
 *	state := ExpandKey (state, 0, salt)
 *      state := ExpandKey(state, 0, password)
 * 4. ctext := "OrpheanBeholderScryDoubt"
 * 5. REPEAT 64:
 * 	ctext := Encrypt_ECB (state, ctext);
 * 6. RETURN Concatenate (salt, ctext);
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "blf.h"
#include "bcrypt.h"

/* This implementation is adaptable to current computing power.
 * You can have up to 2^31 rounds which should be enough for some
 * time to come.
 */

static void encode_salt(char *, uint8_t *, uint16_t, uint8_t);
static void encode_base64(uint8_t *, uint8_t *, uint16_t);
static void decode_base64(uint8_t *, uint16_t, uint8_t *);

static const uint8_t Base64Code[] =
"./ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";

static const uint8_t index_64[128] = {
	255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
	255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
	255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
	255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
	255, 255, 255, 255, 255, 255, 0, 1, 54, 55,
	56, 57, 58, 59, 60, 61, 62, 63, 255, 255,
	255, 255, 255, 255, 255, 2, 3, 4, 5, 6,
	7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
	17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27,
	255, 255, 255, 255, 255, 255, 28, 29, 30,
	31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
	41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
	51, 52, 53, 255, 255, 255, 255, 255
};
#define CHAR64(c)  ( (c) > 127 ? 255 : index_64[(c)])

static void
decode_base64(uint8_t *buffer, uint16_t len, uint8_t *data)
{
	uint8_t *bp = buffer;
	uint8_t *p = data;
	uint8_t c1, c2, c3, c4;
	while (bp < buffer + len) {
		c1 = CHAR64(*p);
		c2 = CHAR64(*(p + 1));

		/* Invalid data */
		if (c1 == 255 || c2 == 255)
			break;

		*bp++ = (c1 << 2) | ((c2 & 0x30) >> 4);
		if (bp >= buffer + len)
			break;

		c3 = CHAR64(*(p + 2));
		if (c3 == 255)
			break;

		*bp++ = ((c2 & 0x0f) << 4) | ((c3 & 0x3c) >> 2);
		if (bp >= buffer + len)
			break;

		c4 = CHAR64(*(p + 3));
		if (c4 == 255)
			break;
		*bp++ = ((c3 & 0x03) << 6) | c4;

		p += 4;
	}
}

static void
encode_salt(char *salt, uint8_t *csalt, uint16_t clen, uint8_t logr)
{
	salt[0] = '$';
	salt[1] = BCRYPT_VERSION;
	salt[2] = 'a';
	salt[3] = '$';

	snprintf(salt + 4, 4, "%2.2u$", logr);

	encode_base64((uint8_t *) salt + 7, csalt, clen);
}
/* Generates a salt for this version of crypt.
   Since versions may change. Keeping this here
   seems sensible.
   output: the computed salt will be stored here. This buffer must be
           at least BCRYPT_SALT_OUTPUT_SIZE bytes. The result will be
           null-terminated.

*/

char   *
ruby_bcrypt_gensalt(uint8_t log_rounds, uint8_t rseed)
{
	char* output = (char*)calloc(sizeof(char), BCRYPT_SALT_OUTPUT_SIZE);
	if (output) {
		if (log_rounds < 4)
			log_rounds = 4;
		else if (log_rounds > 31)
			log_rounds = 31;

		encode_salt(output, &rseed, BCRYPT_MAXSALT, log_rounds);
	}
	return output;
}
/* We handle $Vers$log2(NumRounds)$salt+passwd$
   i.e. $2$04$iwouldntknowwhattosayetKdJ6iFtacBqJdKe6aW7ou */

char   *
ruby_bcrypt(const char *key, const char *salt)
{
	blf_ctx state;
	uint32_t rounds, i, k;
	uint16_t j;
	uint8_t key_len, salt_len, logr, minor;
	uint8_t ciphertext[4 * BCRYPT_BLOCKS] = "OrpheanBeholderScryDoubt";
	uint8_t csalt[BCRYPT_MAXSALT];
	uint32_t cdata[BCRYPT_BLOCKS];
	int n;

	/* Discard "$" identifier */
	salt++;

	if (*salt > BCRYPT_VERSION) {
		return NULL;
	}

	/* Check for minor versions */
	if (salt[1] != '$') {
		 switch (salt[1]) {
		 case 'a':
			 /* 'ab' should not yield the same as 'abab' */
			 minor = salt[1];
			 salt++;
			 break;
		 default:
			 return NULL;
		 }
	} else
		 minor = 0;

	/* Discard version + "$" identifier */
	salt += 2;

	if (salt[2] != '$')
		/* Out of sync with passwd entry */
		return NULL;

	/* Computer power doesn't increase linear, 2^x should be fine */
	n = atoi(salt);
	if (n > 31 || n < 0)
		return NULL;
	logr = (uint8_t)n;
	if ((rounds = (uint32_t) 1 << logr) < BCRYPT_MINROUNDS)
		return NULL;

	/* Discard num rounds + "$" identifier */
	salt += 3;

	if (strlen(salt) * 3 / 4 < BCRYPT_MAXSALT)
		return NULL;

	/* We dont want the base64 salt but the raw data */
	decode_base64(csalt, BCRYPT_MAXSALT, (uint8_t *) salt);
	salt_len = BCRYPT_MAXSALT;
	key_len = strlen(key) + (minor >= 'a' ? 1 : 0);

	/* Setting up S-Boxes and Subkeys */
	Blowfish_initstate(&state);
	Blowfish_expandstate(&state, csalt, salt_len,
	    (uint8_t *) key, key_len);
	for (k = 0; k < rounds; k++) {
		Blowfish_expand0state(&state, (uint8_t *) key, key_len);
		Blowfish_expand0state(&state, csalt, salt_len);
	}

	/* This can be precomputed later */
	j = 0;
	for (i = 0; i < BCRYPT_BLOCKS; i++)
		cdata[i] = Blowfish_stream2word(ciphertext, 4 * BCRYPT_BLOCKS, &j);

	/* Now do the encryption */
	for (k = 0; k < 64; k++)
		blf_enc(&state, cdata, BCRYPT_BLOCKS / 2);

	for (i = 0; i < BCRYPT_BLOCKS; i++) {
		ciphertext[4 * i + 3] = cdata[i] & 0xff;
		cdata[i] = cdata[i] >> 8;
		ciphertext[4 * i + 2] = cdata[i] & 0xff;
		cdata[i] = cdata[i] >> 8;
		ciphertext[4 * i + 1] = cdata[i] & 0xff;
		cdata[i] = cdata[i] >> 8;
		ciphertext[4 * i + 0] = cdata[i] & 0xff;
	}

	char* output = (char*)calloc(sizeof(char), BCRYPT_OUTPUT_SIZE);

	if (output) {
		i = 0;
		output[i++] = '$';
		output[i++] = BCRYPT_VERSION;
		if (minor)
			output[i++] = minor;
		output[i++] = '$';

		snprintf(output + i, 4, "%2.2u$", logr);

		encode_base64((uint8_t *) output + i + 3, csalt, BCRYPT_MAXSALT);
		encode_base64((uint8_t *) output + strlen(output), ciphertext,
		    4 * BCRYPT_BLOCKS - 1);
	}
	return output;
}

static void
encode_base64(uint8_t *buffer, uint8_t *data, uint16_t len)
{
	uint8_t *bp = buffer;
	uint8_t *p = data;
	uint8_t c1, c2;
	while (p < data + len) {
		c1 = *p++;
		*bp++ = Base64Code[(c1 >> 2)];
		c1 = (c1 & 0x03) << 4;
		if (p >= data + len) {
			*bp++ = Base64Code[c1];
			break;
		}
		c2 = *p++;
		c1 |= (c2 >> 4) & 0x0f;
		*bp++ = Base64Code[c1];
		c1 = (c2 & 0x0f) << 2;
		if (p >= data + len) {
			*bp++ = Base64Code[c1];
			break;
		}
		c2 = *p++;
		c1 |= (c2 >> 6) & 0x03;
		*bp++ = Base64Code[c1];
		*bp++ = Base64Code[c2 & 0x3f];
	}
	*bp = '\0';
}
